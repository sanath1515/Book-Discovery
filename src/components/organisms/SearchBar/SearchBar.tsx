import InputField from "../../atoms/InputField";
import { makeStyles, MenuItem, Typography } from "@material-ui/core";
import SearchIcon from "@material-ui/icons/Search";
import React, { useEffect, useState } from "react";
import SearchResultCard from "../../molecules/SearchResultCard/SearchResultCard";
import { COLORS } from "../../../theme/constants";
import ArrowForwardIosIcon from "@material-ui/icons/ArrowForwardIos";
import { bookProp } from "../../../data/type";
import { useDispatch } from "react-redux";
import { useHistory } from "react-router";
import { getAllBooks } from "../../pages/ApiProvider";

const useStyles = makeStyles((theme) => ({
  root: {
    padding: theme.spacing(0.5, 1),
    display: "flex",
    alignItems: "center",
    width: 380,
    margin: "0",
    boxShadow: "none",
    flex: "null",
    border: "1px solid",
    borderColor: COLORS.GREY_300,
    borderRadius: "1px",
    zIndex: 3,
    marginRight:theme.spacing(2)
  },
  icon: {
    marginLeft: theme.spacing(2.5),
    marginRight: theme.spacing(3.5),

    color: COLORS.ICON_500,
  },
  closeIcon: {
    marginLeft: "auto",
    color: COLORS.ICON_500,
    fontSize: "20px",
  },
  flex: {
    flex: "column",
  },
  flexRow: {
    display: "flex",
    padding: theme.spacing(1, 1, 1, 18),
    borderRadius: "4px",
    boxShadow: "none",
    border: "1px solid",
    backgroundColor: "white",
    borderColor: COLORS.GREY_300,
    color: COLORS.PRIMARY_COLOR_500,
    zIndex: 1,
    cursor: "pointer",
  },
  searchBooks: {
    backgroundColor: "white",
    width: "390px",
    position: "relative",
    top: "5px",
    display: "flex",
    flexDirection: "column",
  },
  menuItem: {
    padding: 0,
    zIndex: 1,
  },
  noResults: {
    display: "flex",
    padding: theme.spacing(1, 1, 1, 12),
    borderRadius: "4px",
    boxShadow: "none",
    border: "1px solid",
    borderColor: COLORS.GREY_300,
    backgroundColor: "white",
    zIndex: 1,
  },
  icon1: {
    width: "12px",
    height: "22px",
    color: COLORS.PRIMAY_500,
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(3),

    cursor: "pointer",
  },
}));

export type SearchBarProps = {
  type?: string;
  searchValue?: string;
  placeholder: string;
  className?: string;
};

const SearchBar: React.FC<SearchBarProps> = ({
  placeholder,
  className,
}) => {
  const classes = useStyles();
  const [bookData, setBookData] = useState<bookProp[]>([]);

  const [filterBooks, setFilterBooks] = useState<bookProp[]>([]);

  const [expand, setExpand] = useState(false);

  const [type, setType] = useState("0");

  const dispatch = useDispatch();
  const history = useHistory();

  const check = (tempDataItem: bookProp, str: string) => {
    return (
      tempDataItem.title!.toLowerCase().includes(str.toLowerCase()) ||
      tempDataItem.author!.toLowerCase().includes(str.toLowerCase())
    );
  };

  useEffect(() => {
    getAllBooks((res:bookProp[])=>{setBookData(res)});
  }, []);

  const onChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    console.log(event.target.value);
    if ((event.target.value as string) === "") {
      setExpand(false);
      return;
    }
    let filteredBooks = bookData.filter((dataItem: bookProp) => {
      return check(dataItem, event.target.value as string);
    });
    if (filteredBooks.length > 4) {
      setType("3");
    } else if (filteredBooks.length === 0) {
      setType("1");
    } else {
      setType("2");
    }
    setFilterBooks(filteredBooks);
    setExpand(true);
    dispatch({ type: "SET_SEARCH_TEXT", payload: event.target.value });
  };

  const handleOnClick = () => {
    setExpand(false);
    
    history.push("/search");
  };

  const handleClose =() =>{
    setExpand(false);
  }


  const searchBooks = filterBooks.slice(0, 4).map((book) => {
    return (
      <MenuItem className={classes.menuItem}>
        <SearchResultCard
          id={book.id!}
          authorName={book.author!}
          title={book.title!}
          field={book.category!}
          image={book.bookImage!}
          handleClose ={handleClose}

        />
      </MenuItem>
    );
  });

  const seeResults = (
    <div className={classes.flexRow}>
      <Typography variant="body2" onClick={handleOnClick}>
        See all results
      </Typography>
      <ArrowForwardIosIcon onClick={handleOnClick} className={classes.icon1} />
    </div>
  );

  const noResults = (
    <div className={classes.noResults}>
      <Typography variant="body2">
        No results found. Please try a different search term.

      </Typography>
    </div>
  );

  return (
    <div className={classes.flex}>
      <div className={classes.root}>
        <SearchIcon className={classes.icon} />
        <InputField
          placeholder={placeholder}
          className={className}
          disableUnderline={true}
          data-testid="search-field"
          onChange={onChange}
        />
      </div>

      <div className={classes.searchBooks}>
        {expand && searchBooks}
        {expand && type === "3" && seeResults}
        {expand && type === "1" && noResults}
      </div>
    </div>
  );
};

export default SearchBar;
