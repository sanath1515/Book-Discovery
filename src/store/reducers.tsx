export type bookState = {
    searchText:string,
    category:string,
    singleViewId:string
}

export const initialBookState = {
    searchText:"",
    category:"",
    singleViewId:"1"
}

type Action = {type:string,payload:string}

export const booksReducer = (state:bookState = initialBookState,action:Action)=>
{
    switch(action.type)
    {
        case "SET_CATEGORY":{
            return {...state,category:action.payload};
        }
        case "SET_SEARCH_TEXT":{
            return {...state,searchText:action.payload};
        }
        case "SET_SINGLE_VIEW_ID":{
            return {...state,singleViewId:action.payload};
        }
        default:
            return state;
    }
}