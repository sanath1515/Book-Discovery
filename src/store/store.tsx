import { createStore } from 'redux';
import { booksReducer } from './reducers';

const store = createStore(booksReducer);

export default store;