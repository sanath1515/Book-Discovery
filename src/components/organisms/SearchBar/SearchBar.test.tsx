import { render, fireEvent } from "@testing-library/react";
import SearchBar from "./SearchBar";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import store from "../../../store/store";

jest.mock('../../pages/ApiProvider', () => ({
    ...(jest.requireActual('../../pages/ApiProvider')),
    getBook: jest.fn(),
    getAllBooks: jest.fn().mockImplementation(() => ({ type: '' })),
  }));

describe("Search Bar test", () => {
  const handleChange = jest.fn();

  it("should render input correctly", () => {
    const { container } = render(
      <Router>
        <Provider store={store}>
          <SearchBar placeholder={""} />{" "}
        </Provider>
      </Router>
    );
    expect(container.querySelector("input")).toBeTruthy();
  });

  it("should match the snapshot", () => {
    const { container } = render(
      <Router>
        <Provider store={store}>
          <SearchBar placeholder={""} />{" "}
        </Provider>
      </Router>
    );
    expect(container.querySelector("input")).toMatchSnapshot();
  });

  it("should call handle change when input value is changed", () => {
    const { container } = render(
      <Router>
        <Provider store={store}>
          <SearchBar placeholder={""} />{" "}
        </Provider>
      </Router>
    );
    const input = container.querySelector("input");
    {
      input && fireEvent.change(input, { target: { value: "value" } });
    }
    expect(fireEvent.change(input!, { target: { value: "value" } })).toBeTruthy;
  });
});
