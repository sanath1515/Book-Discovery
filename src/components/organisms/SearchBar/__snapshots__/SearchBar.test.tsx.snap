// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar test should match the snapshot 1`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputTypeSearch MuiInput-inputTypeSearch"
  placeholder=""
  type="search"
  value=""
/>
`;
